### PR TITLE
1275: An unknown sub-command gives a confusing error message

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

When using the unknown command, the client will output the unexpected message. For example:

```
$ git-pr lis 123

error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]
	-h, --help      Show help
	    --verbose   Turn on verbose output
	    --debug     Turn on debugging output
	    --version   Print the version of this tool
```

When the client can't identify the sub-command `git-pr lis`, it will execute the default command `git-pr help` and will pass `lis 123` as the argument to the `git-pr help` command. And the `git-pr help` can resolve only one argument, which is `lis` in this example, so it would output `error: unexpected input: 123`.

This patch will output more information to the user:

```
$ git-pr lis 123

Can't find the sub-command 'lis', the default sub-command 'help' will be executed.
The arguments [lis, 123] will be passed to the default sub-command.

error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]
	-h, --help      Show help
	    --verbose   Turn on verbose output
	    --debug     Turn on debugging output
	    --version   Print the version of this tool
```

The added information can let the user know why the following output is shown.

Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [SKARA-1275](https://bugs.openjdk.java.net/browse/SKARA-1275): An unknown sub-command gives a confusing error message

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/skara pull/1257/head:pull/1257` \
`$ git checkout pull/1257`

Update a local copy of the PR: \
`$ git checkout pull/1257` \
`$ git pull https://git.openjdk.java.net/skara pull/1257/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1257`

View PR using the GUI difftool: \
`$ git pr show -t 1257`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/skara/pull/1257.diff">https://git.openjdk.java.net/skara/pull/1257.diff</a>

</details>

### CODE REVIEW

**Comment by erikj79** [comment]
The new message does explain in great detail how the implementation interprets the given input, but I think it's unnecessarily verbose. I think something like this would be clear enough:

```
$ git-pr lis 123

error: unknown sub-command: lis
error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]
	-h, --help      Show help
	    --verbose   Turn on verbose output
	    --debug     Turn on debugging output
	    --version   Print the version of this tool
```

However, not all users of MultiCommandParser have the help command as default. The webrev command does not and printing all of this text when running `webrev` with no explicit sub-command would be a regression IMO. Printing this error would need to be made conditional on if a sub-command is required or not. For `pr` it is, while for `webrev` it's not.

**Comment by lgxbslgx** [comment]
If we only output `error: unknown sub-command: lis`, the following output is confused:

```
error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]  // <-- here should be `git-pr help [options] [<COMMAND>] `. See SKARA-1274
	-h, --help      Show help
	    --verbose   Turn on verbose output
	    --debug     Turn on debugging output
	    --version   Print the version of this tool
```

The user would confuse why the client outputs the usage of the `git-pr help`. Because the user expects the following message.

```
error: unknown sub-command: lis

Available sub-commands are:
- apply
- approve
- cc
```

So I think it is necessary to let user know what is happening. As for `git webrev`, it will output the following message in my patch. Because the `git-webrev generate` will execute.

```
Can't find the sub-command 'lis', the default sub-command 'generate' will be executed.
The arguments [lis, 123] will be passed to the default sub-command.
```

Now I suggest the following information:

```
$ git-pr lis 123

error: unknown sub-command: lis
the default sub-command 'help' will be executed with the arguments [lis, 123].

error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]
	-h, --help      Show help
	    --verbose   Turn on verbose output
	    --debug     Turn on debugging output
	    --version   Print the version of this tool
```

```
$ git-webrev lis 123

error: unknown sub-command: lis
the default sub-command 'generate' will be executed with the arguments [lis, 123].

error: unexpected input: 123
usage: git webrev [options] [<FILE>]
// ignore
```

**Comment by lgxbslgx** [comment]
Because the code is still running and not stop. It is good to change the `error` to `warning`. The message could be:

```
$ git-pr lis 123

warning: unknown sub-command: lis
the default sub-command 'help' will be executed with the arguments [lis, 123].

error: unexpected input: 123
usage: git-pr [options] [<COMMAND>]
// ignore
```

**Comment by erikj79** [comment]
I like the last suggestion much better, but I still don't think it should be printed for `git webrev`. Running webrev without a sub-command is normal procedure and should not result in a warning.

**Comment by lgxbslgx** [comment]
@erikj79 I updated the code just now. The new code uses the `warning` and excludes the `git webrev` command as you suggested.

**Comment by lgxbslgx** [comment]
I merged the master branch to re-run the pre-submit tests in macOS, which failed in the past. And ping for review. Thanks.

File: args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java [hunk 1]
```diff
@@ -62,6 +62,11 @@ public Executable parse(String[] args) {
                 var forwardedArgs = Arrays.copyOfRange(args, 1, args.length);
                 return () -> p.main(forwardedArgs);
             }
+            if (!"git-webrev".equals(programName) && !"git webrev".equals(programName)) {
```
Comment (erikj79) [inline args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: I was expecting a more elegant solution to this. I would add a property/method on the Command interface that  MutliCommandParser can use to check this. Something like Command::defaultCommandExpected.

Comment (lgxbslgx) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: > I was expecting a more elegant solution to this. I would add a property/method on the Command interface that MutliCommandParser can use to check this. Something like Command::defaultCommandExpected.

Good idea. Should we add it in this patch? Or we should file another issue to implement it?

Comment (erikj79) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: This was meant as a suggestion for this patch. I don't think the current solution is ok to go in.

I'm not sure about the name of the new method, it's hard to come up with a good name for this property of a command. It's trying to express if the default sub-command is expected to be run with arguments or not, or if it takes arguments at all, but putting all that in a method name becomes quite a mouthful.

Comment (lgxbslgx) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: What about `defaultCommandWarningEnabled`? Actually, the default sub-command is always invoked if no sub-command is provided. But the warning message are outputted optionally.

Comment (lgxbslgx) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: What about `defaultCommandWarningEnabled`? Actually, the default sub-command is always invoked if no sub-command is provided. But the warning message are outputted optionally.

Comment (erikj79) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java:65]: Works for me. Thanks.

Comment (lgxbslgx) [reply args/src/main/java/org/openjdk/skara/args/MultiCommandParser.java]: I updated the code just now by using the property `defaultCommandWarningEnabled`.

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**